modules/improvedigitalBidAdapter.js

```diff
@@ -0,0 +1,386 @@
+const LIB_VERSION_GLOBAL = '3.0.5';
+
+var CONSTANTS = require('src/constants');
+var utils = require('src/utils');
+var bidfactory = require('src/bidfactory');
+var bidmanager = require('src/bidmanager');
+var adloader = require('src/adloader');
+var Adapter = require('src/adapter').default;
+var adaptermanager = require('src/adaptermanager');
+
+const IMPROVE_DIGITAL_BIDDER_CODE = 'improvedigital';
+
+const ImproveDigitalAdapter = function () {
+  let baseAdapter = new Adapter(IMPROVE_DIGITAL_BIDDER_CODE);
+  baseAdapter.idClient = new ImproveDigitalAdServerJSClient('hb');
+
+  let LIB_VERSION = LIB_VERSION_GLOBAL;
+
+  // Ad server needs to implement JSONP using this function as the callback
+  let CALLBACK_FUNCTION = '$$PREBID_GLOBAL$$' + '.improveDigitalResponse';
+
+  baseAdapter.getNormalizedBidRequest = function(bid) {
+    let adUnitId = utils.getBidIdParameter('placementCode', bid) || null;
+    let placementId = utils.getBidIdParameter('placementId', bid.params) || null;
+    let publisherId = null;
+    let placementKey = null;
+
+    if (placementId === null) {
+      publisherId = utils.getBidIdParameter('publisherId', bid.params) || null;
+      placementKey = utils.getBidIdParameter('placementKey', bid.params) || null;
+    }
+    let keyValues = utils.getBidIdParameter('keyValues', bid.params) || null;
+    let localSize = utils.getBidIdParameter('size', bid.params) || null;
+    let bidId = utils.getBidIdParameter('bidId', bid);
+
+    let normalizedBidRequest = {};
+    if (placementId) {
+      normalizedBidRequest.placementId = placementId;
+    } else {
+      if (publisherId) {
+        normalizedBidRequest.publisherId = publisherId;
+      }
+      if (placementKey) {
+        normalizedBidRequest.placementKey = placementKey;
+      }
+    }
+
+    if (keyValues) {
+      normalizedBidRequest.keyValues = keyValues;
+    }
+    if (localSize && localSize.w && localSize.h) {
+      normalizedBidRequest.size = {};
+      normalizedBidRequest.size.h = localSize.h;
+      normalizedBidRequest.size.w = localSize.w;
+    }
+    if (bidId) {
+      normalizedBidRequest.id = bidId;
+    }
+    if (adUnitId) {
+      normalizedBidRequest.adUnitId = adUnitId;
+    }
+    return normalizedBidRequest;
+  }
+
+  $$PREBID_GLOBAL$$.improveDigitalResponse = function(response) {
+    let bidRequests = utils.getBidderRequestAllAdUnits(IMPROVE_DIGITAL_BIDDER_CODE);
+    if (bidRequests && bidRequests.bids && bidRequests.bids.length > 0) {
+      utils._each(bidRequests.bids, function (bidRequest) {
+        let bidObjects = response.bid || [];
+        utils._each(bidObjects, function (bidObject) {
+          if (bidObject.id === bidRequest.bidId) {
+            let bid;
+            if (!bidObject.price || bidObject.price === null) {
+              bid = bidfactory.createBid(CONSTANTS.STATUS.NO_BID, bidRequest);
```

Comment:
repeated code blocks should go into a function

Reply:
Fixed.